File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
        if (!root) {
            return !subRoot ? true : false;
        }
        if (subRoot && root->val == subRoot->val && sameTree(root, subRoot)) {
            return true;
        }
        return isSubtree(root->left, subRoot) || isSubtree(root->right, subRoot);
    }

    bool sameTree(TreeNode* root1, TreeNode *root2) {
        if (!root1 && !root2) return true;
        if (!root1 || !root2 || root1->val != root2->val) return false;

        TreeNode *left1 = root1->left, *left2 = root2->left;
        TreeNode *right1 = root1->right, *right2 = root2->right;

        return sameTree(left1, left2) && sameTree(right1, right2);
    }
};

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        maxSum : int = float("-inf")
        
        def dfs(curr):
            nonlocal maxSum
            if not curr: return 0

            left_val = dfs(curr.left)
            right_val = dfs(curr.right)
            max_val = max(left_val, right_val)

            if max_val <= 0: s = curr.val
            else: s = max_val + curr.val

            maxSum = max(maxSum, max(s, left_val + right_val + curr.val))

            return s

        dfs(root)
        return maxSum


            